### ISSUE TITLE
IntelliJ fails to Sync jvmTarget configuration of Kotlin projects

### ISSUE THREAD

**Issue by jjohannes** [gradle/gradle#8244]
This issue is in the latest `5.2` nightly and was introduced with https://github.com/gradle/gradle/commit/fefcd4d30563aa7fcedaba69b6382fc63a6869f7.

**Reproduce**
- Initialize a simple Kotlin project: `gradlew init --dsl kotlin --type kotlin-application`
- Add a `jvmTarge` configuration to the build script
```
tasks.withType<KotlinCompile> {
    kotlinOptions.jvmTarget = "1.8"
}
```
- Import in IntelliJ
- _Sync_ using a nightly **without** the above commit - e.g. `5.2-20181221000038+0000`. The _target platform_ is correct (1.8):
![image](https://user-images.githubusercontent.com/1963746/51211378-99643680-1915-11e9-889c-de949663a367.png)
- _Sync_ with a current nightly **with** the above commit - e.g. `5.2-20190110000045+0000`. The _target platform_ is wrong (1.6):
![screen shot 2019-01-15 at 17 17 27](https://user-images.githubusercontent.com/1963746/51211329-6fab0f80-1915-11e9-9c06-e3e262c3b22a.png)

It needs to be determined if this also requires changes on the IntelliJ side for `5.2`.

**Comment by wolfs** [comment]
@adammurdoch Any idea what could cause this?

**Comment by adammurdoch** [comment]
My guess would be that IDEA is looking for a `_Decorated` in the class name and then doing something special. The generated classes have different names now.

If so, IDEA should use public APIs to do whatever it is doing. The class names are not part of the public API and will change over time as we add different features. Even using some of the internal interfaces and classes would be better than using the class names.

**Comment by JohannesRudolph** [comment]
This appears to be broke in 5.3

```
17:19 $ ./gradlew --version

------------------------------------------------------------
Gradle 5.3
------------------------------------------------------------

Build time:   2019-03-20 11:03:29 UTC
Revision:     f5c64796748a98efdbf6f99f44b6afe08492c2a0
```

**Comment by wolfs** [comment]
@JohannesRudolph Could you elaborate what is broken in 5.3? The current bug has been fixed. Are you saying it re-appeared?

**Comment by JohannesRudolph** [comment]
Yes, we upgraded from gradle 4.x straight to 5.3 and have had this in our build.gradle for a long time 

```groovy
    tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).all {
        sourceCompatibility = 1.8
        targetCompatibility = 1.8

        kotlinOptions {
            allWarningsAsErrors = true // best practice
            jvmTarget = "1.8"
        }
    }
```

but now we see issues related to jvmTarget (e.g. compiler errors re. java static interface method etc.)

**Comment by wolfs** [comment]
@JohannesRudolph I just tried with Gradle 5.3 and IDEA 2019.1 and this seems to work:
```
tasks.withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile>().all {
    kotlinOptions.jvmTarget = "1.8"
}
```

Note that `sourceCompatibility` and `targetCompatibility` is probably not set on the `KotlinCompile` task but on the java extension. So it would be better to add it to top-level in the build file.

**Comment by JohannesRudolph** [comment]
Thanks a lot @wolfs  - I'll have to check if that makes any difference. We're on IDEA 2018.3.6 for now. I think `sourceCompatibility` and `targetCompatibility` may be legacy artifacts of our (long ago successfully completed) java to kotlin migration, will check if we can remove those. 

Do you think it could make any difference at all that we do the kotlinOptions configuration within a `subprojects { subproject -> ... }` block? We didn't have any issues on Gradle 4.2, so this appeared only after updating to 5.3.

**Comment by wolfs** [comment]
No change comes to mind with respect to upgrading. Can you check out the migration guide: https://docs.gradle.org/current/userguide/upgrading_version_4.html#for_all_users ?

Especially try to update to the latest 4.x version (4.10.3) and then try fixing the deprecation warnings before jumping on 5.x. Does the IDE configuration work in 4.10.3?

**Comment by JohannesRudolph** [comment]
I did some more testing and it appears to only affect our own build. I tried messing with https://github.com/gradle/kotlin-dsl/tree/master/samples/multi-kotlin-project (which is the closest sample to our own setup) and IDEA handles  importing `jvmTarget="1.8"` correctly with Gradle 5.3 here. 

I've run the steps to verify we're not using any deprecated features and that hasn't turned up anything useful. So far my team has been able to workaround this issue by manually overriding jvmTarget in IDEA, but we'd like to get to the bottom of the misconfiguration. Something's apparently broken with our gradle files, in gradle or in IDEA.
